Add ability to mask URLs from DOM and JavaScript.

xeenon opened the pull request:
#### c0c652e4bd3ca6542ebe26fcd55d36a47dca3c28
<pre>
Add ability to mask URLs from DOM and JavaScript.
<a href="https://bugs.webkit.org/show_bug.cgi?id=238576">https://bugs.webkit.org/show_bug.cgi?id=238576</a>
rdar://81991245

Reviewed by Chris Dumez.

* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(GetterExpression): Use getURLAttributeForBindings for URL types.
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::jsTestObj_reflectedURLAttrGetter): Updated to use getURLAttributeForBindings.
(WebCore::jsTestObj_reflectedUSVURLAttrGetter): Ditto.
(WebCore::jsTestObj_reflectedCustomURLAttrGetter): Ditto.
* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::value const): Use getAttributeForBindings.
* Source/WebCore/dom/Attr.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::shouldMaskURLForBindings const): Added.
(WebCore::Document::hasURLsToMaskForBindings const): Added.
(WebCore::Document::maskedURLForBindingsIfNeeded const): Added.
(WebCore::Document::maskedURLStringForBindings const): Added.
(WebCore::Document::maskedURLForBindings const): Added.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::getAttribute const): Use getAttributeInternal for the Attribute lookup.
(WebCore::Element::getAttributeForBindings const): Added.
(WebCore::Element::innerHTML const): Use new ResolveURLs::NoExcludingURLsForPrivacy when calling serializeFragment.
(WebCore::Element::outerHTML const): Ditto.
(WebCore::Element::resolveURLStringIfNeeded const): Added. Moved from MarkupAccumulator::resolveURLIfNeeded and added new NoExcludingURLsForPrivacy path.
Changed YesExcludingLocalFileURLsForPrivacy to YesExcludingURLsForPrivacy and included shouldMaskURLForBindings() check.
(WebCore::Element::completeURLsInAttributeValue const): Use resolveURLStringIfNeeded instead of making URL directly.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.idl: Use getAttributeForBindings for getAttribute, and getAttributeNSForBindings for getAttributeNS.
* Source/WebCore/dom/ElementInlines.h:
(WebCore::Element::getURLAttributeForBindings const): Added.
(WebCore::Element::getAttributeInternal const): Added. Copied from getAttribute to share Attribute lookup code.
(WebCore::Element::getAttributeNSForBindings const): Added.
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::innerHTML const): Use new ResolveURLs::NoExcludingURLsForPrivacy when calling serializeFragment.
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::resolveURLIfNeeded const): Use Element::resolveURLStringIfNeeded.
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::selectionInHTMLFormat): Changed YesExcludingLocalFileURLsForPrivacy to YesExcludingURLsForPrivacy.
* Source/WebCore/editing/gtk/EditorGtk.cpp:
(WebCore::Editor::writeSelectionToPasteboard): Ditto.
* Source/WebCore/editing/libwpe/EditorLibWPE.cpp:
(WebCore::Editor::writeSelectionToPasteboard): Ditto.
* Source/WebCore/editing/mac/EditorMac.mm:
(WebCore::Editor::writeImageToPasteboard): Ditto.
* Source/WebCore/editing/markup.cpp:
(WebCore::sanitizedMarkupForFragmentInDocument): Ditto.
* Source/WebCore/editing/markup.h:
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::completeURLsInAttributeValue const): Use resolveURLStringIfNeeded instead of making URL directly.
(WebCore::HTMLImageElement::srcsetForBindings const):Added. Use getAttributeForBindings with ResolveURLs::YesExcludingURLsForPrivacy.
(WebCore::HTMLImageElement::setSrcsetForBindings): Added.
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLImageElement.idl:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::shouldMaskURLForBindings const): Added.
* Source/WebCore/page/Page.h:
(WebCore::Page::hasURLsToMaskForBindings const):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/platform/win/PasteboardWin.cpp:
(WebCore::Pasteboard::writeSelection): Changed YesExcludingLocalFileURLsForPrivacy to YesExcludingURLsForPrivacy.
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const): Added maskedURLSchemes.
(WebKit::WebPageCreationParameters::decode): Ditto.
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::copy const): Added maskedURLSchemes.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::maskedURLSchemes const): Added.
(API::PageConfiguration::setMaskedURLSchemes): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _maskedURLSchemes]): Added.
(-[WKWebViewConfiguration _setMaskedURLSchemes:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters): Added maskedURLSchemes.
* Source/WebKit/WebProcess/WebCoreSupport/gtk/WebEditorClientGtk.cpp:
(WebKit::WebEditorClient::updateGlobalSelection): Changed YesExcludingLocalFileURLsForPrivacy to YesExcludingURLsForPrivacy.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::WebPage): Added maskedURLSchemes.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewConfiguration.mm:
(TEST(WebKit, ConfigurationMaskedURLSchemes)): Added.

Canonical link: <a href="https://commits.webkit.org/251962@main">https://commits.webkit.org/251962@main</a>
</pre>

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5646,6 +5646,41 @@ URL Document::completeURL(const String& url, ForceUTF8 forceUTF8) const
+    return page->hasURLsToMaskForBindings();
+}
+
+URL Document::maskedURLForBindingsIfNeeded(const URL& url) const
```
Why can't this return a `const URL&`?

cdumez commented on Source/WebCore/dom/ElementInlines.h:
```diff
@@ -90,6 +90,11 @@ inline const AtomString& Element::attributeWithoutSynchronization(const Qualifie
     return nullAtom();
 }
 
+inline URL Element::getURLAttributeForBindings(const QualifiedName& name) const
```
Can this return a `const URL&` ?

xeenon replied:
This needed to be `URL` to work. This is what was causing the 30+ failures in `dom/html/level2/html` and it was crashing on macOS 13.

```
Exception Type:        EXC_BAD_ACCESS (SIGKILL)
Exception Codes:       KERN_INVALID_ADDRESS at 0xef3b80014a3cef54 -> 0x000000014a3cef54 (possible pointer authentication failure)
Exception Codes:       0x0000000000000001, 0xef3b80014a3cef54

Termination Reason:    Namespace PAC_EXCEPTION, Code 1 

VM Region Info: 0x14a3cef54 is in 0x1493e0000-0x150998000;  bytes after start: 16707412  bytes before end: 106729643
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      MALLOC_SMALL                143800000-144000000    [ 8192K] rw-/rwx SM=PRV  
      GAP OF 0x53e0000 BYTES
--->  __TEXT                      1493e0000-150998000    [117.7M] r-x/rwx SM=COW  ...ons/A/WebCore
      __DATA_CONST                150998000-150c2c000    [ 2640K] r--/rw- SM=COW  ...ons/A/WebCore

Application Specific Information:
dyld config: DYLD_LIBRARY_PATH=/Volumes/Timothy/Work/Safari/OpenSource/WebKitBuild/Debug DYLD_FRAMEWORK_PATH=/Volumes/Timothy/Work/Safari/OpenSource/WebKitBuild/Debug
CRASHING TEST: dom/html/level2/html/HTMLFrameElement08.html


Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   WebCore                       	       0x1499e0ddc WTF::StringImpl::length() const + 12 (StringImpl.h:298)
1   WebCore                       	       0x14aace66c JSC::jsStringWithCache(JSC::VM&, WTF::String const&) + 60 (JSString.h:1032)
2   WebCore                       	       0x149a619a8 WebCore::JSConverter<WebCore::IDLUSVString>::convert(JSC::JSGlobalObject&, WTF::String const&) + 40 (JSDOMConvertStrings.h:132)
3   WebCore                       	       0x14a291dcc WebCore::JSConverter<WebCore::IDLAtomStringAdaptor<WebCore::IDLUSVString> >::convert(JSC::JSGlobalObject&, WTF::URL const&) + 52 (JSDOMConvertStrings.h:251)
4   WebCore                       	       0x14a291d84 JSC::JSValue WebCore::JSConverterOverloader<WebCore::IDLAtomStringAdaptor<WebCore::IDLUSVString>, true, false>::convert<WTF::URL const&>(JSC::JSGlobalObject&, WTF::URL const&) + 52 (JSDOMConvertBase.h:117)
5   WebCore                       	       0x14a291d3c JSC::JSValue WebCore::toJS<WebCore::IDLAtomStringAdaptor<WebCore::IDLUSVString>, WTF::URL const&>(JSC::JSGlobalObject&, WTF::URL const&) + 52 (JSDOMConvertBase.h:146)
6   WebCore                       	       0x14a3cef50 JSC::JSValue WebCore::toJS<WebCore::IDLAtomStringAdaptor<WebCore::IDLUSVString>, WTF::URL const&>(JSC::JSGlobalObject&, JSC::ThrowScope&, WTF::URL const&) + 56 (JSDOMConvertBase.h:185)
7   WebCore                       	       0x14a2c5198 WebCore::jsHTMLFrameElement_srcGetter(JSC::JSGlobalObject&, WebCore::JSHTMLFrameElement&) + 180 (JSHTMLFrameElement.cpp:279)
8   WebCore                       	       0x14a20aa50 long long WebCore::IDLAttribute<WebCore::JSHTMLFrameElement>::get<&(WebCore::jsHTMLFrameElement_srcGetter(JSC::JSGlobalObject&, WebCore::JSHTMLFrameElement&)), (WebCore::CastedThisErrorBehavior)3>(JSC::JSGlobalObject&, long long, JSC::PropertyName) + 288 (JSDOMAttribute.h:88)
```

cdumez reviewed (approved):
LGTM

xeenon committed:
Add ability to mask URLs from DOM and JavaScript.
https://bugs.webkit.org/show_bug.cgi?id=238576
rdar://81991245

Reviewed by Chris Dumez.

* Source/WebCore/bindings/scripts/CodeGenerator.pm:
(GetterExpression): Use getURLAttributeForBindings for URL types.
* Source/WebCore/bindings/scripts/test/JS/JSTestObj.cpp:
(WebCore::jsTestObj_reflectedURLAttrGetter): Updated to use getURLAttributeForBindings.
(WebCore::jsTestObj_reflectedUSVURLAttrGetter): Ditto.
(WebCore::jsTestObj_reflectedCustomURLAttrGetter): Ditto.
* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::value const): Use getAttributeForBindings.
* Source/WebCore/dom/Attr.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::shouldMaskURLForBindings const): Added.
(WebCore::Document::hasURLsToMaskForBindings const): Added.
(WebCore::Document::maskedURLForBindingsIfNeeded const): Added.
(WebCore::Document::maskedURLStringForBindings const): Added.
(WebCore::Document::maskedURLForBindings const): Added.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::getAttribute const): Use getAttributeInternal for the Attribute lookup.
(WebCore::Element::getAttributeForBindings const): Added.
(WebCore::Element::innerHTML const): Use new ResolveURLs::NoExcludingURLsForPrivacy when calling serializeFragment.
(WebCore::Element::outerHTML const): Ditto.
(WebCore::Element::resolveURLStringIfNeeded const): Added. Moved from MarkupAccumulator::resolveURLIfNeeded and added new NoExcludingURLsForPrivacy path.
Changed YesExcludingLocalFileURLsForPrivacy to YesExcludingURLsForPrivacy and included shouldMaskURLForBindings() check.
(WebCore::Element::completeURLsInAttributeValue const): Use resolveURLStringIfNeeded instead of making URL directly.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.idl: Use getAttributeForBindings for getAttribute, and getAttributeNSForBindings for getAttributeNS.
* Source/WebCore/dom/ElementInlines.h:
(WebCore::Element::getURLAttributeForBindings const): Added.
(WebCore::Element::getAttributeInternal const): Added. Copied from getAttribute to share Attribute lookup code.
(WebCore::Element::getAttributeNSForBindings const): Added.
* Source/WebCore/dom/ShadowRoot.cpp:
(WebCore::ShadowRoot::innerHTML const): Use new ResolveURLs::NoExcludingURLsForPrivacy when calling serializeFragment.
* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::MarkupAccumulator::resolveURLIfNeeded const): Use Element::resolveURLStringIfNeeded.
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::selectionInHTMLFormat): Changed YesExcludingLocalFileURLsForPrivacy to YesExcludingURLsForPrivacy.
* Source/WebCore/editing/gtk/EditorGtk.cpp:
(WebCore::Editor::writeSelectionToPasteboard): Ditto.
* Source/WebCore/editing/libwpe/EditorLibWPE.cpp:
(WebCore::Editor::writeSelectionToPasteboard): Ditto.
* Source/WebCore/editing/mac/EditorMac.mm:
(WebCore::Editor::writeImageToPasteboard): Ditto.
* Source/WebCore/editing/markup.cpp:
(WebCore::sanitizedMarkupForFragmentInDocument): Ditto.
* Source/WebCore/editing/markup.h:
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::completeURLsInAttributeValue const): Use resolveURLStringIfNeeded instead of making URL directly.
(WebCore::HTMLImageElement::srcsetForBindings const):Added. Use getAttributeForBindings with ResolveURLs::YesExcludingURLsForPrivacy.
(WebCore::HTMLImageElement::setSrcsetForBindings): Added.
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLImageElement.idl:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::shouldMaskURLForBindings const): Added.
* Source/WebCore/page/Page.h:
(WebCore::Page::hasURLsToMaskForBindings const):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/platform/win/PasteboardWin.cpp:
(WebCore::Pasteboard::writeSelection): Changed YesExcludingLocalFileURLsForPrivacy to YesExcludingURLsForPrivacy.
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const): Added maskedURLSchemes.
(WebKit::WebPageCreationParameters::decode): Ditto.
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::copy const): Added maskedURLSchemes.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::maskedURLSchemes const): Added.
(API::PageConfiguration::setMaskedURLSchemes): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _maskedURLSchemes]): Added.
(-[WKWebViewConfiguration _setMaskedURLSchemes:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters): Added maskedURLSchemes.
* Source/WebKit/WebProcess/WebCoreSupport/gtk/WebEditorClientGtk.cpp:
(WebKit::WebEditorClient::updateGlobalSelection): Changed YesExcludingLocalFileURLsForPrivacy to YesExcludingURLsForPrivacy.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::WebPage): Added maskedURLSchemes.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewConfiguration.mm:
(TEST(WebKit, ConfigurationMaskedURLSchemes)): Added.

Canonical link: https://commits.webkit.org/251962@main
